### PR TITLE
Pileup-based read error corrector

### LINKED ISSUE #4868: Somatic error correction for low allele fractions
Read error correction tends to follow the basic strategy of 1) collect kmer counts 2) replace rare kmers with their closest non-rare match.  For germline calling where there is a huge gap between error rates and diploid het allele fractions this is sufficient.  Mutect, however, must contend with cases where counts alone do not discriminate perfectly between errors and real mutations.

Without committing to an approach, it seems like phasing might help.  That is, we could construct haplotypes of rare kmers and error correct those.  This should work because sequencing errors are unphased and real variants are.  There are phased artifacts, of course, but we handle those in downstream filtering.

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #4868.

@takutosato This reduces false positives and improves speed at no cost to sensitivity.  I'm not quite ready to turn it on by default but I want it in the code to begin experiments, such as combining with linked de Bruijn graphs and FFPE error correction.

### CODE REVIEW

**Comment by takutosato** [comment]
@davidbenjamin Can I assume that you didn't change `NearbyKmerErrorCorrector` (other than renaming and refactoring)

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                offset = 4; // just some randomly circulating offset for error position
+        }
+
+        // now correct all reads
+        final ReadErrorCorrector readErrorCorrector = new PileupReadErrorCorrector(3.0, header);
+        readErrorCorrector.correctReads(finalizedReadList);
+
+        // check that corrected reads have exactly same content as original reads
+        for (int k=0; k < numBadReads; k++) {
+            final byte[] badBases = finalizedReadList.get(k).getBases();
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java]: pretty sure we need `finalizedReadList.get(numGoodReads+k).getBases();` here. And when I change it the tests fail—the bad reads are not corrected.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        int offset = 0;
+        final byte[] quals = new byte[readLength];
+
+        Arrays.fill(quals,(byte)30);
+
+        for (int k=0; k < numGoodReads; k++) {
+            final byte[] bases = Arrays.copyOfRange(refChunk.getBytes(),offset,offset+readLength);
+            final GATKRead read = ArtificialReadUtils.createArtificialRead(header, bases, quals,readLength+"M");
+            read.setName("good_read" + k);
+            read.setReadGroup(rg.getReadGroupId());
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java]: I'm surprised that LocusIteratorByState works without start positions, let alone unsorted.

Comment (takutosato) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java]: I think all of these start at the same position (1:10000). I think you intended to slide the start position by `offset`. I think adding the start positions to each read should do the trick (and then sort before you pass the list to LocusIteratorByState).

**Comment by davidbenjamin** [comment]
> @davidbenjamin Can I assume that you didn't change NearbyKmerErrorCorrector (other than renaming and refactoring)

Yes you can.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    // and we don't correct bases
+    private static final int INDEL_SPAN = 15;
+    private static final int INDEL_MISMATCHES = 3;
+
+    public PileupReadErrorCorrector(final double logOddsThreshold, final SAMFileHeader header) {
+        this.logOddsThreshold = logOddsThreshold;
+        this.header = header;
+    }
+
+    public final List<GATKRead> correctReads(final Collection<GATKRead> originalRreads) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java (hunk 2 above)
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java]: done + done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java (hunk 1 above)
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrectorUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            }
+
+            int lastEdit = size - 1;
+            for (int n = size - 1; n >= INDEL_MISMATCHES - 1 && edits.get(n).getLeft() - edits.get(n - INDEL_MISMATCHES + 1).getLeft()  < INDEL_SPAN; n--) {
+                lastEdit = n - INDEL_MISMATCHES;
+            }
+
+            for (int n = firstEdit; n <= lastEdit; n++) {
+                bases[edits.get(n).getLeft()] = edits.get(n).getRight();
+                quals[edits.get(n).getLeft()] = GOOD_QUAL;
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java]: It does now.  Totally rewrote the test.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PileupReadErrorCorrector.java]: done

**Comment by davidbenjamin** [comment]
back to @takutosato

**Pull request merged**